### PR TITLE
[next-v5.0] Refactor ERC721 _transfer function to support minting and burning

### PR DESCRIPTION
**Pull request by Amxx**
Part of https://github.com/OpenZeppelin/openzeppelin-contracts/issues/3535

Also remove before and after transfer hooks

Replace #3635

#### PR Checklist

- [ ] Tests
- [ ] Documentation
- [ ] Changelog entry

### CODE REVIEW

File: contracts/token/ERC721/ERC721.sol [hunk 1]
```diff
@@ -234,103 +241,11 @@ contract ERC721 is Context, ERC165, IERC721, IERC721Metadata {
...
-     * WARNING: Usage of this method is discouraged, use {_safeMint} whenever possible
-     *
-     * Requirements:
-     *
-     * - `tokenId` must not exist.
-     * - `to` cannot be the zero address.
-     *
-     * Emits a {Transfer} event.
-     */
-    function _mint(address to, uint256 tokenId) internal virtual {
```
Comment (frangio) [inline contracts/token/ERC721/ERC721.sol:279]: I don't think we should remove `_mint`, `_burn`, etc. It's a lot nicer and logical to write `_mint(...)` than `_transfer(address(0), ...)`.

I realize this introduces serious concerns with overrides so these shorthand functions should not be virtual and people should be directed to override `_transfer` instead.

**Comment by frangio** [comment]
We should do the same with ERC20.

File: contracts/token/ERC721/ERC721.sol (hunk 1 above)
Comment (Amxx) [reply contracts/token/ERC721/ERC721.sol:279]: +1 for the "keep them in a non virtual way"

**Comment by Amxx** [comment]
> We should do the same with ERC20 and 1155.

yes, but in other PRs :)

File: contracts/token/ERC721/extensions/ERC721URIStorage.sol [hunk 2]
```diff
@@ -48,12 +48,16 @@ abstract contract ERC721URIStorage is ERC721 {
...
-    function _burn(uint256 tokenId) internal virtual override {
-        super._burn(tokenId);
+    function _transfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal virtual override {
+        super._transfer(from, to, tokenId);
 
         if (bytes(_tokenURIs[tokenId]).length != 0) {
```
Comment (frangio) [inline contracts/token/ERC721/extensions/ERC721URIStorage.sol:62]: This function should only do this if `to` is 0!

**Comment by frangio** [comment]
Superceded by https://github.com/OpenZeppelin/openzeppelin-contracts/pull/3893.

**Pull request closed without merging**